rmw_fastrtps_dynamic_cpp/src/type_support_registry.cpp

```diff
@@ -0,0 +1,159 @@
+        auto members = static_cast<const rosidl_typesupport_introspection_cpp::MessageMembers *>(
+          ros_type_support->data);
+        return new MessageTypeSupport_cpp(members, ros_type_support);
+      }
+      RMW_SET_ERROR_MSG("Unknown typesupport identifier");
+      return nullptr;
+    };
+
+  return get_type_support(ros_type_support, message_types_, creator_fun);
+}
+
+type_support_ptr TypeSupportRegistry::get_request_type_support(
+  const rosidl_service_type_support_t * ros_type_support)
+{
+  auto creator_fun = [&ros_type_support]() -> type_support_ptr
+    {
+      if (using_introspection_c_typesupport(ros_type_support->typesupport_identifier)) {
+        auto members = static_cast<const rosidl_typesupport_introspection_c__ServiceMembers *>(
+          ros_type_support->data);
+        return new RequestTypeSupport_c(members, ros_type_support);
+      } else if (using_introspection_cpp_typesupport(ros_type_support->typesupport_identifier)) {
+        auto members = static_cast<const rosidl_typesupport_introspection_cpp::ServiceMembers *>(
+          ros_type_support->data);
+        return new RequestTypeSupport_cpp(members, ros_type_support);
+      }
+      RMW_SET_ERROR_MSG("Unknown typesupport identifier");
+      return nullptr;
+    };
+
+  return get_type_support(ros_type_support, request_types_, creator_fun);
+}
+
+type_support_ptr TypeSupportRegistry::get_response_type_support(
+  const rosidl_service_type_support_t * ros_type_support)
+{
+  auto creator_fun = [&ros_type_support]() -> type_support_ptr
+    {
+      if (using_introspection_c_typesupport(ros_type_support->typesupport_identifier)) {
+        auto members = static_cast<const rosidl_typesupport_introspection_c__ServiceMembers *>(
+          ros_type_support->data);
+        return new ResponseTypeSupport_c(members, ros_type_support);
+      } else if (using_introspection_cpp_typesupport(ros_type_support->typesupport_identifier)) {
+        auto members = static_cast<const rosidl_typesupport_introspection_cpp::ServiceMembers *>(
+          ros_type_support->data);
+        return new ResponseTypeSupport_cpp(members, ros_type_support);
+      }
+      RMW_SET_ERROR_MSG("Unknown typesupport identifier");
+      return nullptr;
+    };
+
+  return get_type_support(ros_type_support, response_types_, creator_fun);
+}
+
+void TypeSupportRegistry::return_message_type_support(
+  const rosidl_message_type_support_t * ros_type_support)
+{
+  return_type_support(ros_type_support, message_types_);
+}
+
+void TypeSupportRegistry::return_request_type_support(
+  const rosidl_service_type_support_t * ros_type_support)
+{
+  return_type_support(ros_type_support, request_types_);
+}
+
+void TypeSupportRegistry::return_response_type_support(
+  const rosidl_service_type_support_t * ros_type_support)
+{
+  return_type_support(ros_type_support, response_types_);
+}
+
+static TypeSupportRegistry type_registry_instance;
```

Comment:
@MiguelCompany consider moving this into the function below to avoid having a non trivially constructed object in global scope.